mlx_lm/quant/dwq.py

```diff
@@ -18,19 +19,62 @@
 from mlx_lm.tuner.utils import print_trainable_parameters
 from mlx_lm.utils import (
     load,
+    load_tokenizer,
+    pipeline_load,
     quantize_model,
     save,
 )
 
 
+def compute_dwq_targets(
+    model,
+    save_dir,
+    train_data,
+    valid_data,
+    batch_size,
+    max_seq_length,
+    seed,
+):
+    rank = mx.distributed.init().rank()
+
+    def _compute_targets(data, path, split):
+
+        if rank == 0:
+            path = path / split
+            path.mkdir(parents=True, exist_ok=True)
+        for i, (batch, _) in (
+            pbar := tqdm(
+                enumerate(iterate_batches(data, batch_size, max_seq_length, seed=seed)),
+                total=len(data) // batch_size,
+                desc=f"Computing targets for {split}",
+                disable=rank != 0,
+            )
+        ):
+            batch = batch[:, :-1]
+            logits = model(batch)
+            # Hack to make the last op pre-eval on the CPU to avoid even timeout
+            logits = mx.stop_gradient(logits, stream=mx.cpu)
```

Comment:
This is an unfortunate hack. Basically if you have the last op on the GPU then the shared event that is on the GPU can timeout waiting for another rank to finish. A possible fix would be to have an atomic event for Metal the same way we do with the Fence.